src/Streamly/Internal/Unicode/Utf8/Type.hs

```diff
@@ -0,0 +1,422 @@
+-- |
+-- Module      : Streamly.Internal.Unicode.Utf8.Type
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+module Streamly.Internal.Unicode.Utf8.Type
+    (
+    -- * Type
+      Utf8
+    , toArray
+
+    -- * Creation and elimination
+    , empty
+    , singleton
+    , toStream
+    , fromStream
+    , pack
+    , unpack
+
+    -- * Basic interface
+    , cons
+    , snoc
+    , append
+    , uncons
+    , unsnoc
+    , head
+    , last
+    , tail
+    , init
+    , null
+
+    , isSingleton
+    , length
+    , compareLength
+
+    -- * Folds
+    , write
+
+    -- * Unfolds
+    , read
+    )
+where
+
+
+#include "inline.hs"
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Control.DeepSeq (NFData)
+import Control.Monad.IO.Class (MonadIO)
+import Data.Bifunctor (second)
+import Data.Bits (shiftR, (.&.))
+import Data.Char (ord)
+import Data.String (IsString(..))
+import Data.Word (Word8)
+import GHC.Base (assert)
+import GHC.IO.Encoding.Failure (isSurrogate)
+import Streamly.Internal.Data.Array.Foreign.Type (Array)
+import Streamly.Internal.Data.Fold (Fold)
+import Streamly.Internal.Data.Stream.IsStream (SerialT)
+import Streamly.Internal.Data.Unfold (Unfold)
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Streamly.Internal.Data.Array.Foreign as Array
+import qualified Streamly.Internal.Data.Array.Foreign.Type as Array
+import qualified Streamly.Internal.Data.Array.Foreign.Mut.Type as MArray
+import qualified Streamly.Internal.Data.Fold.Type as Fold
+import qualified Streamly.Internal.Data.Stream.IsStream as Stream
+import qualified Streamly.Internal.Unicode.Stream as Unicode
+
+import Prelude hiding (head, init, last, length, null, read, tail)
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+-- >>> import qualified Streamly.Internal.Unicode.Utf8 as Utf8
+
+--------------------------------------------------------------------------------
+-- Type
+--------------------------------------------------------------------------------
+
+-- | A space efficient, packed, unboxed Unicode container.
+newtype Utf8 =
+    Utf8 (Array Word8)
+    deriving (NFData)
+
+empty :: Utf8
+empty = Utf8 Array.nil
+
+--------------------------------------------------------------------------------
+-- Helpers
+--------------------------------------------------------------------------------
+
+{-# INLINE toStream #-}
+toStream :: Monad m => Utf8 -> SerialT m Char
+toStream = Unicode.decodeUtf8 . Array.toStream . toArray
+
+-- XXX Try removing IO
+-- XXX This would require rewriting Array.Foreign.write-ish functions. We can't
+-- use the underlying Array.Foreign.Mut.write-ish
+{-# INLINE fromStream #-}
+fromStream :: SerialT IO Char -> Utf8
```

Comment:
This must be `fromStream :: SerialT m Char -> m Utf8`. You cannot have the result pure when the stream is not pure.